### PR TITLE
Query: add query metrics to calls going through the Store API

### PR DESCRIPTION
**Pull request by douglascamata**
* [x] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

## Changes

Overall, this pull request's intent is to implement the first step in the proposal [More granular query performance metrics](https://github.com/thanos-io/thanos/blob/main/docs/proposals-accepted/202108-more-granular-query-performance-metrics.md).

## Verification

- Built the containers and ran it locally.
- Added some tests, including e2e tests.

### CODE REVIEW

File: pkg/store/metrics/metrics.go
Comment (jessicalins) [inline pkg/store/metrics/metrics.go:16]: is `SeriesQueryPerformanceMetricsAggregator` being used outside the `metrics` pkg? If not, I think we could leave it unexported and use the exported constructor `NewSeriesQueryPerformanceMetricsAggregator`

File: cmd/thanos/query.go
Comment (jessicalins) [inline cmd/thanos/query.go:695]: I wonder if we could shorten the name a bit (e.g. to `NewSeriesQueryMetricsAggregator`) - since from the params being used we can infer that it aggregate performance metrics - wdyt? :)

File: pkg/query/querier.go
Comment (jessicalins) [inline pkg/query/querier.go:32]: same for `SeriesStatsReporter` - do we need it exported?

File: pkg/query/querier.go
Comment (jessicalins) [inline pkg/query/querier.go:51]: does `QueryableCreator` needs to be exported as well?

File: pkg/query/querier.go
Comment (jessicalins) [inline pkg/query/querier.go:34]: and `NoopSeriesStatsReporter`?

File: pkg/query/querier.go
Comment (jessicalins) [inline pkg/query/querier.go:64]: and `NewQueryableCreator`?

File: cmd/thanos/query.go
Comment (jessicalins) [reply cmd/thanos/query.go:695]: also for some of the other names that include `Performance`

**Review by jessicalins — COMMENTED** [review]
looking good! :star2: 
I left some questions/suggestions

**Comment by fpetkovski** [comment]
Should we start documenting metrics somewhere in our docs?

File: cmd/thanos/query.go
Comment (douglascamata) [reply cmd/thanos/query.go:695]: I'm all in for a shorter name that will still be descriptive. Couldn't find myself. I think your suggestion is good! Might even call it just `metricsm.NewQueryPerformanceAggregator`. What do you think, @jessicalins?

**Comment by douglascamata** [comment]
@fpetkovski I believe we should. One of the first things that looked for when I started to learn more Thanos was a list of all metrics it exports and their meaning. I still find myself very often digging through code to figure out what metrics there are, what they measure, etc.

It'll be quite some work, but super useful!

File: pkg/query/querier.go
Comment (douglascamata) [reply pkg/query/querier.go:51]: Yes yes, this was already exported because it used in other packages.

File: pkg/query/querier.go
Comment (douglascamata) [reply pkg/query/querier.go:34]: It is used in other packages.

**Comment by fpetkovski** [comment]
We can maybe start with these ones that you added now, and extend the list over time?

File: pkg/store/metrics/metrics.go
Comment (bwplotka) [inline pkg/store/metrics/metrics.go:4]: Metrics or telemetry? Let's be consistent (: (with the flag name)

File: pkg/store/metrics/metrics.go
Comment (bwplotka) [inline pkg/store/metrics/metrics.go:60]: I know we agreed in proposal to this, but can we double check how to use this metric? How to visualizate this in Grafana/Prometheus? Can we make short experiment. I think it's quite a precedence here - I don't think `histogram_quantile` would work with this 🤔

**Review by bwplotka — COMMENTED** [review]
Amazing work! Generally looks good, just two questions. 👍🏽

File: pkg/store/metrics/metrics.go
Comment (douglascamata) [reply pkg/store/metrics/metrics.go:60]: We can visualize it with queries like so:

```
histogram_quantile(0.95,
    sum(rate(
        thanos_store_api_query_duration_seconds{samples_le=1000, series_le=10}[5m]
    )) by (le)
)
```

As long as we don't mix different histograms, as in `histogram_quantile(0.95, sum(rate(thanos_store_api_query_duration_seconds{samples_le=~"1000|10000", series_le=~"10|100"}[5m])) by (le))`, we won't have issues with mixed up bucketing and the queries will be fine.

Comment (douglascamata) [reply pkg/store/metrics/metrics.go]: Actually it would work fine even on `histogram_quantile(0.95, sum(rate(thanos_store_api_query_duration_seconds{samples_le=~"1000|10000", series_le=~"10|100"}[5m])) by (le))`, because the buckets are the same. 

We can aggregate visualize it analogous to this example from Prometheus' documentation: `histogram_quantile(0.9, sum by (job, le) (rate(http_request_duration_seconds_bucket[10m])))`:

```
histogram_quantile(0.95,
    sum(rate(
        thanos_store_api_query_duration_seconds{samples_le=1000, series_le=10}[5m]
    )) by (samples_le, series_le, le)
)
```

Comment (douglascamata) [reply pkg/store/metrics/metrics.go]: > I don't think histogram_quantile would work with this 🤔

@bwplotka why do you think so? I am not able to see why it wouldn't work, but I don't have as much experience as you on this.

File: docs/components/query.md
Comment (matej-g) [inline docs/components/query.md:477]: May I request adding an issue for this? :innocent: I think this would be good first issue for someone.

Comment (douglascamata) [reply docs/components/query.md:477]: Done: #5758.

**Review by jessicalins — DISMISSED** [review]
:rocket:

File: pkg/store/telemetry/telemetry.go
Comment (matej-g) [inline pkg/store/telemetry/telemetry.go:4]: I would still advise against breaking this into another package unless there's a reason.

Comment (bwplotka) [reply pkg/store/telemetry/telemetry.go:4]: I don't mind personally any of those.

File: pkg/store/telemetry/telemetry.go
Comment (bwplotka) [inline pkg/store/telemetry/telemetry.go:60]: My main concerns is still this: Is such "weird" metric usable in PromQL? (two dimensional histogram)

**Review by bwplotka — APPROVED** [review]
LGTM, mod one concern (:

File: pkg/store/telemetry/telemetry.go
Comment (bwplotka) [reply pkg/store/telemetry/telemetry.go:60]: Discussed in https://github.com/thanos-io/thanos/pull/5741#discussion_r987038615

Comment (bwplotka) [reply pkg/store/telemetry/telemetry.go:60]: OK, I am fine then, thanks for explanation (:

**Review by bwplotka — DISMISSED** [review]

File: pkg/store/telemetry/telemetry.go
Comment (douglascamata) [reply pkg/store/telemetry/telemetry.go]: @matej-g I got rid of the package. Could you re-review the PR, please? 🙏

**Review by matej-g — DISMISSED** [review]

**Review by matej-g — APPROVED** [review]
(previously approved, approving after merging `main` and merging this PR now)

File: pkg/store/telemetry.go
Comment (christopherzli) [inline pkg/store/telemetry.go:45]: doesn't number of series overlap too?

Comment (douglascamata) [reply pkg/store/telemetry.go:45]: It does, but we are (or were back then) not tracking the what individual series was touched. It would be quite expensive to do the tracking.

File: pkg/api/query/grpc.go
Comment (christopherzli) [inline pkg/api/query/grpc.go:172]: why is this set to no-op here?

Comment (douglascamata) [reply pkg/api/query/grpc.go:172]: Because this is an internally initiated request, not something started by a user.

**Comment by christopherzli** [comment]
just curious if we have thinking about expose these stats with the query response as well?

**Pull request merged**
